### PR TITLE
make @clz() @ctz() @popCount() @byteSwap() @bitReverse() all take types

### PR DESCRIPTION
**Pull request by shawnl**
breaking: make @clz() @ctz() @popCount() @bSwap() @bitReverse() all take types
    
    Closes: #2119
    
    breaking: @bitreverse -> @bitReverse, @bswap -> @bSwap
    
    Closes: #2120
    
    Comes with tests

### CODE REVIEW

File: std/special/compiler_rt.zig [hunk 1]
```diff
@@ -9,7 +9,7 @@ comptime {
     @export("__getf2", @import("compiler_rt/comparetf2.zig").__getf2, linkage);
 
     if (!is_test) {
-        // only create these aliases when not testing
+        // Why don't we need these when testing?
```
Comment (andrewrk) [inline std/special/compiler_rt.zig]: I don't remember, to be honest. You can delete my unhelpful comment (apologies for that). I don't think changing it to a question helps anything though.

File: src/ir.cpp [hunk 2]
```diff
@@ -16844,92 +16828,145 @@ static IrInstruction *ir_analyze_instruction_optional_unwrap_ptr(IrAnalyze *ira,
...
-                    value->value.type->data.integral.bit_count);
-            IrInstruction *result = ir_const(ira, &clz_instruction->base, return_type);
-            bigint_init_unsigned(&result->value.data.x_bigint, result_usize);
-            return result;
-        }
 
-        IrInstruction *result = ir_build_clz(&ira->new_irb,
-            clz_instruction->base.scope, clz_instruction->base.source_node, value);
-        result->value.type = return_type;
+    if (op->value.type->id == ZigTypeIdComptimeInt) {
```
Comment (andrewrk) [inline src/ir.cpp]: Your proposal #2121 was rejected. This pull request implements #2121 as if it were accepted. 

What are you expecting to accomplish by doing this?

File: std/special/compiler_rt.zig (hunk 1 above)
Comment (shawnl) [reply std/special/compiler_rt.zig]: I figured it out. These are aliases, not the canonical names...I will update this comment

File: src/ir.cpp (hunk 2 above)
Comment (shawnl) [reply src/ir.cpp]: Well, it *does* implicitly cast, but yes this check should be removed too. (ever occurrence)

**Comment by shawnl** [comment]
1 sec, bugs

File: src/analyze.cpp [hunk 3]
```diff
@@ -339,6 +339,19 @@ ZigType *get_smallest_unsigned_int_type(CodeGen *g, uint64_t x) {
...
+static uint8_t bits_needed_for_no_zero_representation_unsigned(uint64_t x) {
+    uint8_t count = 0;
+    for (uint64_t s = x - 1;s != 0;s >>= 1)
+        count++;
+
+    return count;
+}
+
+//Equivilent to math.Log2Int. ceil log2int
+ZigType *get_shift_type(CodeGen *g, uint64_t x) {
```
[the remark this answers is not in the document]
Comment (shawnl) [reply src/analyze.cpp]: can't. There are differn't. I added a comment explaining _why_.

**Comment by shawnl** [comment]
needs some more tests.

**Comment by andrewrk** [comment]
still working on this?

**Comment by andrewrk** [comment]
Merged in e09c05f689bbc5c34e7e7cde89bbf0d28ab56421

**Pull request closed without merging**